Add tempel-clojure receipe

mpenet opened the pull request:
### Brief summary of what the package does

This adds support for clojure using tempel "snippets".  

### Direct link to the package repository

https://github.com/mpenet/tempel-clojure

### Your association with the package

Maintainer 

### Relevant communications with the upstream package maintainer

I used clojure-snippets as a base, testing via straight.el the package is working.

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [] My elisp byte-compiles cleanly
- [] `M-x checkdoc` is happy with my docstrings
- [] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

minad commented:
See also #8285. There is also a discussion on the [Tempel issue tracker](https://github.com/minad/tempel/issues/68) if it is better to package templates separately per language or as a single collection package.

riscy commented:
@minad you wrote --

> On the other hand it could happen also the other way round. If https://github.com/melpa/melpa/pull/8282 gets accepted into MELPA, the MELPA maintainers may reject a catch all template package and instead suggest that per-language packages are the way to go.

I think it may be good if MELPA can (try) to adhere to the convention preferred by the tempel maintainer(s) -- which seems to be one megapack? -- but some of the MELPA maintainers (i.e. me) can be forgetful.  From memory I think I've tried to coax yasnipper contributors into contributing very small snippet packages to larger collections, or to start their own larger collection that invites more contributions.  For example instead of "Ronda's Ruby snippets", I'd suggest to Ronda to consider becoming the steward of a "Rubby snippets" megapack that I could direct future people to.

@mpenet as for the code itself, very quick first pass --

`tempel-clojure.el` with checkdoc 0.6.2:
```
tempel-clojure.el:1: White space found at end of line
```

`tempel-clojure.el` with package-lint 20221003.1636:
```
1 issue found:
55:1: warning: `eval-after-load' is for use in configurations, and should rarely be used in packages.
```

Package and license:
- Prefer the default recipe or `:defaults`, if possible. (e.g. `(tempel-clojure :repo "mpenet/tempel-clojure" :fetcher github :files (:defaults "templates"))`
- tempel-clojure-pkg.el -- consider excluding; MELPA creates one

mpenet commented:
@riscy thanks for the review. 

I updated the recipe as you suggested, I also fixed the whitespace warning and killed the pkg file as you suggested. 
The autoload is intentional, I think that makes sense to keep it for such a package. 

On the issue mentioned in the tempel repo I exposed my opinion on package vs megapack for these things already. I don't want to bore you with that here again. I will leave this PR open and let you and @minad decide what you think is best. 
If a megapack ends up being the final choice I think that it would be better if the repo is under @minad's org/account at lesat, even if external contributors are added.

minad commented:
On 11/25/22 20:15, Chris Rayner wrote:
> I think it may be good if MELPA can (try) to adhere to the convention
> preferred by the tempel maintainer(s) -- which seems to be one megapack?
> -- but some of the MELPA maintainers (i.e. me) can be forgetful. From
> memory I think I've tried to coax yasnipper contributors into
> contributing very small snippet packages to larger collections, or to
> start their own larger collection that invites more contributions. For
> example instead of "Ronda's Ruby snippets", I'd suggest to Ronda to
> consider becoming the steward of a "Rubby snippets" megapack that I
> could direct future people to.

I don't have a clear preference except that it would be nice if there is
an agreement on a convention. Having both a collection and individual
packages does not seem like a good outcome. But I don't want to judge
and I don't want to enforce a convention for now. We can just continue
with both packaging conventions and maybe in the future there is some
convergence and packages may get obsolete.

------------------------------------------------------------------------

Disadvantages of individual packages:

- More work for the users, the package maintainers, the ELPA/MELPA
maintainers, ...
- Packages may get unmaintained more easily
- Code duplication (Template registration code)

Advantages of individual packages:

- There are no unnecessary templates around
- Fine-grained control for the user
- Slightly more efficient since Tempel loads all templates on the path
at once. But we could probably optimize that such that Tempel only keeps
the templates in memory which are needed by current major modes.
- More control for the individual maintainers

Disadvantages of a single package:

- Disagreements of maintainers
- Single control instance? Maybe not if there are multiple maintainers?
- Bloat?

Advantages of a single package:

- Less work for all the involved people
- A single place for users and contributors
- No code duplication

minad commented:
@mpenet It seems to me that having a single package with multiple maintainers may be slightly advantageous. However I don't want to force anyone to follow some convention, and you seem to have strong opinion against a single archive.

If there is a decision on a single archive, I can create the repository in an Org or on my account and add you, @Crandel and maybe more interested people as maintainers. But to be clear, I don't have the bandwidth to maintain it. I also don't want to be the one who has to make administrative decisions - this will be up to the team of maintainers.

Otherwise we could just wait and see if people are converging to a convention.

mpenet commented:
On the other hand, there could be a mix of both solutions without duplication:

a megapack that would contain what is supported out of the box by emacs, or what is considered "defaults" (fundamental, lisp, org & co). And single packages for more exotic things, like specific programming languages (clojure, ruby, haskell, etc).

minad commented:
> On the other hand, there could be a mix of both solutions without duplication:
> a megapack that would contain what is supported out of the box by emacs, or what is considered "defaults" (fundamental, lisp, org & co). And single packages for more exotic things, like specific programming languages (clojure, ruby, haskell, etc).

@mpenet This is something which would happen eventually. However I don't agree that there is a clear line where the cut could be made. None of Clojure, Ruby, Haskell is exotic. Ruby is supported by Emacs out of the box. Using Emacs builtin support as criterion is not a good idea, since there are many obscure prog modes provided by Emacs.

mpenet commented:
@minad 

Like I mentioned it's mostly because I lived the same situation with yas, where I maintained a single package for a language, then a mega pack came out, decided to include it and now we have 2 packages that evolve in parallel with different features/bugfixes. One has a very quick turnaround for fixes the other does not and then you have people who hit issues with which one takes over depending on what is installed first/last. But maybe it's more an organisational issue than anything else.

mpenet commented:
> @mpenet This is something which would happen eventually. However I don't agree that there is a clear line where the cut could be made. None of Clojure, Ruby, Haskell is exotic. Ruby is supported by Emacs out of the box. Using Emacs builtin support as criterion is not a good idea, since there are many obscure prog modes provided by Emacs.

That's true, that crossed my mind while I wrote this, that's why I mentioned "what is considered "defaults"".

minad commented:
@mpenet Well, but what is the problem then to do it differently this time if we just start with a megapack right away? Note that the "mega"pack was there first, except that it is not yet "mega", @Crandel didn't really push it and didn't try to get it into MELPA/ELPA.

minad commented:
> That's true, that crossed my mind while I wrote this, that's why I mentioned "what is considered "defaults"".

There is no such thing as defaults, this would be a pointless definition. I have many language modes in my Emacs and I use many.

mpenet commented:
@minad I don't want want to die on that hill :) I have opinions and I understand not everybody has to share them on these things. I am not fundamentally against a megapack, I think it has to be done right however.

mpenet commented:
> If there is a decision on a single archive, I can create the repository in an Org or on my account and add you, @Crandel and maybe more interested people as maintainers. But to be clear, I don't have the bandwidth to maintain it. I also don't want to be the one who has to make administrative decisions - this will be up to the team of maintainers.

That seem like a reasonable proposal.

minad commented:
Okay, let's wait for a few days for @Crandel to chime in and to hear some other opinions. 

For a potential megapack one thing should probably resolved however. Instead of loading all the templates at once, one should load individual files/directories per mode to speed up startup. The loader tempel-path-templates should be replaced or modified. The loader itself could also be part of the megapack.

https://github.com/minad/tempel/blob/d4ffafa6beb895eae0f4d39e525fdd340ca8b505/tempel.el#L443-L465

On the other hand we could also just wait until loading becomes a problem. Since we load sexps, loading is fast. :shrug:

Crandel commented:
Hi @minad, @mpenet,
I created `tempel-collections`, because there were no such package, as yas replacement, at all. I open to share maintenance with any volunteers and maintainers. @mpenet if you want I can add you as maintainer, so you could add and/or update clojure or any other collection as well.    

For now `tempel-collections` not a megapack at all. It has only 3 general collections and one program specific. So I would say it has no problem with speed, but if @minad provide different API for mode specific loads, I will use it. I'm not the best list developer, so it would be hard for me to do the loader myself.

minad commented:
Okay, I think it is easiest if we collaborate on tempel-collection then. @Crandel could you please add @mpenet and myself to the repository? If more people contribute support for more modes they should also be added as additional maintainers. Many people have commit access to Emacs and that also works somehow. We will see how it goes...

Crandel commented:
Invites were sent

mpenet commented:
Great! thanks. I am super busy these next few/past days, but I'll look into opening a PR with the clojure templates.
